lib/src/code_field/models/code_model.dart

```diff
@@ -0,0 +1,103 @@
+import 'package:flutter/material.dart';
+
+import 'code_line_model.dart';
+
+class CodeModel {
+  final String code;
+  late final List<CodeLineModel> lines;
+  late final List<CodeLineModel> readOnlyLines;
+
+  static final _cache = {};
+
+  factory CodeModel({required String code}) {
+    return _cache.putIfAbsent(code, () => CodeModel._(code));
+  }
+
+  CodeModel._(this.code) {
+    lines = splitTextToLines(code);
+    readOnlyLines = getReadOnlyLines(lines);
+  }
+
+  List<CodeLineModel> splitTextToLines(String text) {
+    int positionIndex = 0;
+    int lineCounter = 0;
+    List<String> codeLines = [];
+    List<CodeLineModel> lines = [];
+
+    if (text.isNotEmpty) {
+      codeLines = text.split('\n');
+
+      for (String line in codeLines) {
+        lineCounter++;
+        if (lines.isNotEmpty) {
```

Comment:
This can be done without special-casing some iterations to make code more clear and robust. If the first line's start position must be 1, `positionIndex` should be initialized with 1.